draft-ietf-quic-tls.md

```diff
@@ -1290,40 +1290,38 @@ Endpoints MUST implement and use the replay protections described in {{!TLS13}},
 however it is recognized that these protections are imperfect.  Therefore,
 additional consideration of the risk of replay are needed.
 
-QUIC is not inherently vulnerable to replay attack.  The management of QUIC
-protocol state based on the frame types defined in {{QUIC-TRANSPORT}} is not
-vulnerable to replay.  Processing of QUIC frames is idempotent and cannot result
-in invalid connection states if frames are reordered or lost.  QUIC connections
-do not produce effects that last beyond the lifetime of the connection, except
-for those produced by the application protocol that QUIC serves.
-
-However, this does not count for costs that endpoints might incur as a result of
-accepting 0-RTT.  A server that accepts 0-RTT is exposed to the cost of handling
-a new connection, plus the cost of processing 0-RTT packets.  If replay
-protections are unable to prevent multiple connections from being initiated,
-this could increase these costs because attackers can send copies of 0-RTT
-packets to different server instances, causing the processing to be repeated.
-Servers MUST ensure that they account for any increase in costs before accepting
-connections or 0-RTT.
+QUIC is not vulnerable to replay attack, except via the application protocol
+information it might carry.  The management of QUIC protocol state based on the
+frame types defined in {{QUIC-TRANSPORT}} is not vulnerable to replay.
+Processing of QUIC frames is idempotent and cannot result in invalid connection
+states if frames are replayed, reordered or lost.  QUIC connections do not
+produce effects that last beyond the lifetime of the connection, except for
+those produced by the application protocol that QUIC serves.
+
+Note:
+
+: TLS session tickets and address validation tokens are used to carry QUIC
+  configuration information between connections.  These MUST NOT be used to
+  carry application state.  The potential for reuse of these tokens means that
+  they require stronger protections against replay.
+
+A server that accepts 0-RTT on a connection incurs a higher cost than accepting
+a connection without 0-RTT.  This includes higher processing and computation
+costs.  Servers need to consider the probability of replay and all associated
+costs when accepting 0-RTT.
```

Comment:
I’m not sure 0-RTT server cost is higher, except relative to the client?

Reply:
Your point being that a client can keep their costs low?  That's not relevant here.  What matters is what the server spends relative to other connections.

Keep in mind that a client can replay a ClientHello to a server and have it do the work of connection establishment (which is considerable in terms of computation).  The delta for 0-RTT is all the processing that might be triggered by the replayed 0-RTT packets, which might mean buffering or application-level actions.  If that work is significant, and it could be, then you have an exposure.  Add the relatively negligible work of accepting packets and decrypting them if you must.

In terms of outright expense to a server, it is entirely possible that spamming bogus ClientHello messages (which don't even need valid key shares, possibly not even valid anything) could more effective than spamming 0-RTT.

Reply:
With that argument a client might as well use 1-RTT connections and make the server pay for Diffie-Helmann exchanges in addition to buffering. What makes 0-RTT costly is the ease with which the server can be made to pay that cost. 0-RTT is expensive because it is cheap for the client to incur th expense on the server.